### PR TITLE
Finish Session-04/task-1

### PR DESCRIPTION
**Pull request by NicatBabayev**

### CODE REVIEW

File: session-4/TaskManagementSystem/main.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+func isNearlyFinished(tasksCompleted int) (int, bool) {
+	var isNearlyFinished bool
+	if tasksCompleted > 90 && tasksCompleted != 100 {
+		isNearlyFinished = true
+	} else {
+		isNearlyFinished = false
+	}
+
+	return tasksRemaining(tasksCompleted), isNearlyFinished
+}
```
Comment (zaur-adigozalov) [inline session-4/TaskManagementSystem/main.go]: TasksAvailable is a const and you can create a function to calculate tasks remaining and whether the project is near completion (multiple return values) in the same function
```suggestion
func calculateTasksRemaining(tasksCompleted int) (int, bool) {
    tasksRemaining := tasksAvailable - tasksCompleted
    nearCompletion := tasksCompleted > 90
    return tasksRemaining, nearCompletion
}
```

File: session-4/TaskManagementSystem/main.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+func isCompleted(tasksCompleted, tasksAvailable int) bool {
+	if tasksCompleted == tasksAvailable {
+		return true
+	} else {
+		return false
+	}
+}
+func tasksRemaining(tasksCompleted int) int {
+	return tasksAvailable - tasksCompleted
+}
```
Comment (zaur-adigozalov) [inline session-4/TaskManagementSystem/main.go:79]: You can remove this method, merging with calculateTasksRemaining

File: session-4/TaskManagementSystem/main.go [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+// Didn't understand the exact instructions for variadic function.
+func printTaskDetails(taskDetails ...map[int]string) {
+	fmt.Println()
+	for x := range taskDetails {
+		for k, v := range taskDetails[x] {
+			fmt.Printf("  Name of the task %d is: %s.\n", k, v)
+		}
+	}
+}
```
Comment (zaur-adigozalov) [inline session-4/TaskManagementSystem/main.go]: It's good to use map, even any type, but we can still use string slices (we used it before in our samples)
```suggestion

    tasks := []string{"Task 1", "Task 2", "Task 3", "Task 4", "Task 5"}
    printTasks(tasks...)
    
    // Variadic function to print any number of tasks
func printTasks(tasks ...string) {
    fmt.Println("Task list:")
    for i, task := range tasks {
        fmt.Printf("%d: %s\n", i+1, task)
    }
}
```

File: session-4/TaskManagementSystem/main.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		m := make(map[int]string)
+		m[1] = "One"
+		m[2] = "Two"
+		m[3] = "Three"
+		m[4] = "Four"
+		m[5] = "Five"
+		m[6] = "Six"
+		m[7] = "Seven"
+
+		printTaskDetails(m)
```
Comment (zaur-adigozalov) [inline session-4/TaskManagementSystem/main.go]: No need to use map for this case, but I like your approach

File: session-4/TaskManagementSystem/main.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	case tasksCompleted > 60:
+		projectProgress = "Final phase"
+	}
+
+	fmt.Println("  Welcome to the Task Management System!")
+	fmt.Println("  Project start date is: 2024-09-18 00:00:00 +0000 UTC")
+	fmt.Println("  Project: Task Management System")
+
+	fmt.Println()
+	if tasksRemaining(tasksCompleted) >= 0 {
```
Comment (zaur-adigozalov) [inline session-4/TaskManagementSystem/main.go]: Still you can use suggested "calculateTasksRemaining" with multiple values

**Review by zaur-adigozalov — CHANGES_REQUESTED** [review]
It works, you did some extra which gives positive points. But still code is not clean, and it's hard to read for other developers

I uploaded the homework task which I'd like to do as homework - https://github.com/zaur-adigozalov/mastering-golang-sessions/blob/main/sessions/4/HOMEWORK.go

You can check it and find some tricks based on homework requirements

**Comment by NicatBabayev** [comment]
Thank you for the comments. I see that I have to practice on clean code principles. Is there any resource you can suggest to learn the basics of clean code? Or some repos with neatly written go code?

**Comment by NicatBabayev** [comment]
I do not want just blindly commit your suggestions. So this PR will be open until I fully understand how refactor my code correctly. If you see that I am missing session-04 in my `main` branch please ignore.

**Pull request merged**